File: elixir/wordy/wordy.exs
defmodule Wordy do
  require Integer

  @doc """
  Calculate the math problem in the sentence.
  """
  @spec answer(String.t) :: integer
  def answer(question) do
    parts = question
    |> valid_start!
    |> parse
    |> List.flatten
    |> valid_parts_count!

    [head|tail] = parts
    do_math(tail, String.to_integer(head))
  end

  defp valid_start!(question) do
    unless String.starts_with?(question, "What is ") do
      raise ArgumentError
    end
    question
  end

  defp valid_parts_count!(parts) do
    c = Enum.count(parts)
    unless c > 2 && Integer.is_odd(c) do
      raise ArgumentError
    end
    parts
  end

  defp parse(question) do
    Regex.scan(~r/\-?\d+|plus|minus|divided by|multiplied by/, question)
  end

  defp do_math([], acc), do: acc
  defp do_math([operator, i|tail], acc) do
    i = String.to_integer(i)
    acc = case operator do
      "plus"          -> acc + i
      "minus"         -> acc - i
      "divided by"    -> acc / i
      "multiplied by" -> acc * i
    end
    do_math(tail, acc)
  end
end

File: typescript/wordy/wordy.ts
class WordProblem {
  private question: string;

  constructor(question: string) {
    this.question = question;
  }

  answer(): number {
    let parts = this.parse();
    if (parts.length < 3 || parts.length % 2 == 0) {
      throw new ArgumentError;
    }
    let out = parseInt(parts[0]);
    for (var x=1; x<parts.length; x=x+2) {
      let [op, r]: string[] = parts.slice(x, x+2);
      const rn = parseInt(r, 10);
      switch (op) {
        case 'plus'          : out += rn; break;
        case 'minus'         : out -= rn; break;
        case 'multiplied by' : out *= rn; break;
        case 'divided by'    : out /= rn; break;
      }
    }
    return out;
  }

  parse(): string[] {
    const r = /\-?\d+|plus|minus|divided by|multiplied by/g;
    let matches = [];
    while (true) {
      let match = r.exec(this.question);
      if (match == null) { break; }
      matches.push(match[0]);
    }
    return matches;
  }
}

class ArgumentError {}

module.exports = {
  WordProblem,
  ArgumentError
};